runtime/dart_vm_initializer.cc

```diff
@@ -9,11 +9,74 @@
 #include "flutter/fml/logging.h"
 #include "flutter/fml/synchronization/shared_mutex.h"
 #include "flutter/fml/trace_event.h"
+#include "flutter/lib/ui/ui_dart_state.h"
+#include "flutter/lib/ui/window/platform_configuration.h"
+#include "third_party/tonic/converter/dart_converter.h"
+#include "third_party/tonic/logging/dart_error.h"
 
+namespace {
 // Tracks whether Dart has been initialized and if it is safe to call Dart
 // APIs.
 static std::atomic<bool> gDartInitialized;
 
+void LogUnhandledException(Dart_Handle exception_handle,
+                           Dart_Handle stack_trace_handle) {
+  const std::string error =
+      tonic::StdStringFromDart(Dart_ToString(exception_handle));
+  const std::string stack_trace =
+      tonic::StdStringFromDart(Dart_ToString(stack_trace_handle));
+
+  auto state = flutter::UIDartState::Current();
+  if (state && state->unhandled_exception_callback()) {
+    auto callback = state->unhandled_exception_callback();
+    if (callback(error, stack_trace)) {
+      return;
+    }
+  }
+
+  // Either the exception handler was not set or it could not handle the
+  // error, just log the exception.
+  FML_LOG(ERROR) << "Unhandled Exception: " << error << std::endl
+                 << stack_trace;
+}
+
+void ReportUnhandledException(Dart_Handle exception_handle,
+                              Dart_Handle stack_trace_handle) {
+  // Hooks.dart will call the error handler on PlatformDispatcher if it is
+  // not null. If it is null, returns false, fall into the !handled branch
+  // below and log.
+  // If it is not null, defer to the return value of that closure
+  // to determine whether to report via logging.
+  bool handled = false;
+  auto state = flutter::UIDartState::Current();
```

Comment:
I tried testing a scenario that executes this during isolate initialization and found that this is still unsafe.  `DartState::Current` currently assumes that `Dart_CurrentIsolateData()` is non-null.

Add a check for this to https://github.com/flutter/engine/blob/main/third_party/tonic/dart_state.cc#L59 that returns null if `Dart_CurrentIsolateData()` is null

Reply:
Done and added a test to tonic that creates an isolate group with nullptr data, asserts `tonic::DartState::Current()` returns nullptr and doesn't segfault in this case.